### PR TITLE
📖 book: add ipam contract

### PR DESCRIPTION
**Pull request by schrej**
**What this PR does / why we need it**:
Adds the [IPAM contract](https://github.com/kubernetes-sigs/cluster-api/blob/main/docs/proposals/20220125-ipam-integration.md) to the book. 

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:

/area ipam
/area documentation

### CODE REVIEW

**Comment by sbueringer** [comment]
@schrej Still interested in this one?

**Comment by schrej** [comment]
Yes, just forgot about it... I think I put it on hold due to some uncertainty with regards to `clusterctl move`. And now I also need to update regarding `spec.clusterName` instead of the annotation.

**Comment by enxebre** [comment]
@schrej is this still wip? lgtm as good starting point

**Comment by sbueringer** [comment]
@schrej friendly reminder 😀

**Comment by schrej** [comment]
I think this should be ready now. I've updated it to include the new `spec.clusterName` field and noted that the cluster name label is deprecated.

Sorry that it took so long.

**Comment by sbueringer** [comment]
@lubronzhan Do you maybe have some time to take a first look? (in case you're familiar with what we did in CAPV at the time)

File: docs/book/src/developer/providers/ipam.md [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+
+Note that the IPAM contract is single-stack. If you need both v4 and v6 addresses, two pools and two IPAddressClaims are necessary.
+
+</aside>
+
+## Data Types
+
+An IPAM provider must define one or more API types for IP address pools. The types:
+
+`1. Must belong to an API group served by the Kubernetes apiserver
```
Comment (lubronzhan) [inline docs/book/src/developer/providers/ipam.md:19]: Formatting looks oof, maybe ``` is not needed?

File: docs/book/src/developer/providers/ipam.md [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+## Data Types
+
+An IPAM provider must define one or more API types for IP address pools. The types:
+
+`1. Must belong to an API group served by the Kubernetes apiserver 
+2. Must be implemented as a CustomResourceDefinition. 
+   The CRD name must have the format produced by `sigs.k8s.io/cluster-api/util/contract.CalculateCRDName(Group, Kind)`. 
+3. Must have the standard Kubernetes "type metadata" and "object metadata"
+`4. Should have a status.conditions field with the following:
```
Comment (lubronzhan) [inline docs/book/src/developer/providers/ipam.md:23]: Same as above

File: docs/book/src/developer/providers/ipam.md [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+Note that the IPAM contract is single-stack. If you need both v4 and v6 addresses, two pools and two IPAddressClaims are necessary.
+
+</aside>
+
+1. Create an IPAddressClaim
+   1. The `spec.poolRef` must reference the pool you want to use
+   2. It should have an owner reference to the infrastructure Machine it is created for (required to support `clusterctl move`)
+   3. It's `spec.clusterName` field should be set (or it should have a `cluster.x-k8s.io/cluster-name` label)
+   4. Ideally it's name is derived from the infrastructure Machine's name
+2. Wait until an IP is allocated, ideally by watching the IPAdressClaim and waiting for `status.addressRef` to be set.
```
Comment (lubronzhan) [inline docs/book/src/developer/providers/ipam.md:78]:
```suggestion
2. Wait until an IP is allocated, ideally by watching the IPAddressClaim and waiting for `status.addressRef` to be set.
```

**Review by lubronzhan — APPROVED** [review]
Sorry for the late reply. Overall LGTM!

File: docs/book/src/developer/providers/ipam.md [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+## Data Types
+
+An IPAM provider must define one or more API types for IP address pools. The types:
+
+`1. Must belong to an API group served by the Kubernetes apiserver 
+2. Must be implemented as a CustomResourceDefinition. 
+   The CRD name must have the format produced by `sigs.k8s.io/cluster-api/util/contract.CalculateCRDName(Group, Kind)`. 
+3. Must have the standard Kubernetes "type metadata" and "object metadata"
+`4. Should have a status.conditions field with the following:
+   1. A Ready condition to represent the overall operational state of the component. It can be based on the summary of more detailed conditions existing on the same object, e.g. instanceReady, SecurityGroupsReady conditions.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:24]: Wondering how this works in the context of https://github.com/kubernetes-sigs/cluster-api/pull/10897 (the migration to metav1.Condition)

File: docs/book/src/developer/providers/ipam.md [hunk 5]
```diff
@@ -0,0 +1,81 @@
...
+### IPAM Provider
+
+An IPAM provider must watch for new, updated and deleted IPAddressClaims that reference an IP address pool that is manged by the provider in their `spec.poolRef` field.
+
+#### Normal IPAddressClaim
+
+1. If the IPAddressClaim does not reference a pool managed by the provider in it's `spec.poolRef`, abort the reconciliation.
+2. If the related Cluster is paused, abort reconciliation
+   1. The related Cluster is referenced using the `spec.clusterName` field or a `cluster.x-k8s.io/cluster-name` label (the latter is deprecated).
+   2. If the field is empty and the annotation is not present, reconciliation can continue.
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:39]: I think this should be label? (if this references the cluster-name label and not the paused annotation, but let's be precise here)

File: docs/book/src/developer/providers/ipam.md [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+   3. If the referenced cluster is not found, abort reconciliation.
+   4. If the referenced cluster has `spec.paused` set or a `cluster.x-k8s.io/paused` annotation, skip reconciliation
+3. Add any required provider-specific Finalziers (you probably need one)
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:42]:
```suggestion
3. Add any required provider-specific finalizers (you probably need one)
```

File: docs/book/src/developer/providers/ipam.md [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+4. Allocate an IP address for the claim
+5. Create an IPAddress object
+   1. It should have the same name as the claim.
+   2. It must have a owner reference with `controller: true` and `blockOwnerDeletion: true` to the Claim
+   3. It must have a owner reference with `controller: false` and `blockOwnerDeletion: true` to the referenced Pool
+   4. It should have a Finalizer that prevents accidental deletion, e.g. `ipam.cluster.x-k8s.io/protect-address`.
+6. Set the `status.addressRef` to the created IPAddress
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:49]:
```suggestion
6. Set the `status.addressRef` on the IPAddressClaim to the created IPAddress
```

File: docs/book/src/developer/providers/ipam.md [hunk 8]
```diff
@@ -0,0 +1,81 @@
...
+
+#### Deleted IPAddressClaim
+
+1. If the related Cluster is paused, abort reconciliation (see 2. above)
+2. Deallocate the IP address
+3. Delete the IPAddress object
+   1. Remove any Finalizers that were set to prevent deletion first
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:56]: What do you mean with "prevent deletion first", more specifically what is the "first" referring to?
(I would probably just drop first)

File: docs/book/src/developer/providers/ipam.md [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+
+<aside class="note">
+
+Note that the IPAM contract is single-stack. If you need both v4 and v6 addresses, two pools and two IPAddressClaims are necessary.
+
+</aside>
+
+1. Create an IPAddressClaim
+   1. The `spec.poolRef` must reference the pool you want to use
+   2. It should have an owner reference to the infrastructure Machine it is created for (required to support `clusterctl move`)
```
Comment (sbueringer) [inline docs/book/src/developer/providers/ipam.md:75]: What about controller / blockOwnerDeletion here?

**Review by sbueringer — COMMENTED** [review]
Thx! Just a few small findings.

File: docs/book/src/developer/providers/ipam.md (hunk 5 above)
Comment (schrej) [reply docs/book/src/developer/providers/ipam.md:39]: I meant the `cluster.x-k8s.io/paused` annotation here, I've made it more clear.

File: docs/book/src/developer/providers/ipam.md [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+An IPAM provider must watch for new, updated and deleted IPAddressClaims that reference an IP address pool that is manged by the provider in their `spec.poolRef` field.
+
+#### Normal IPAddressClaim
+
+1. If the IPAddressClaim does not reference a pool managed by the provider in it's `spec.poolRef`, abort the reconciliation.
+2. If the related Cluster is paused, abort reconciliation
+   1. The related Cluster is referenced using the `spec.clusterName` field or a `cluster.x-k8s.io/cluster-name` label (the latter is deprecated).
+   2. If the field is empty and the annotation is not present, reconciliation can continue.
+   3. If the referenced cluster is not found, abort reconciliation.
+   4. If the referenced cluster has `spec.paused` set or a `cluster.x-k8s.io/paused` annotation, skip reconciliation
```
[the remark this answers is not in the document]
Comment (schrej) [reply docs/book/src/developer/providers/ipam.md:41]: It probably makes sense to drop the check for the annotation then.

File: docs/book/src/developer/providers/ipam.md (hunk 9 above)
Comment (schrej) [reply docs/book/src/developer/providers/ipam.md:75]: Good point. We're currently investigating an issue when deleting metal3 based clusters where claims are deadlocked due to the cluster vanishing before they are cleaned up. The paused check prevents the claim from being released if the cluster can't be found. We'll either have to make sure that the cluster doesn't get deleted, or ignore the paused check when releasing addresses.

File: docs/book/src/developer/providers/ipam.md [hunk 11]
```diff
@@ -0,0 +1,81 @@
...
+3. Add any required provider-specific Finalziers (you probably need one)
+4. Allocate an IP address for the claim
+5. Create an IPAddress object
+   1. It should have the same name as the claim.
+   2. It must have a owner reference with `controller: true` and `blockOwnerDeletion: true` to the Claim
+   3. It must have a owner reference with `controller: false` and `blockOwnerDeletion: true` to the referenced Pool
+   4. It should have a Finalizer that prevents accidental deletion, e.g. `ipam.cluster.x-k8s.io/protect-address`.
```
[the remark this answers is not in the document]
Comment (schrej) [reply docs/book/src/developer/providers/ipam.md:48]: The generic reconciler that we've implemented in the in-cluster provider uses this annotation. It probably makes sense to just match the domain with whatever is used for the CRDs.

File: docs/book/src/developer/providers/ipam.md (hunk 4 above)
Comment (fabriziopandini) [reply docs/book/src/developer/providers/ipam.md:24]: I'm updating contracts to take into account changes introduced by #10897, I will take care of this one too after this PR merge

File: docs/book/src/developer/providers/ipam.md (hunk 9 above)
Comment (schrej) [reply docs/book/src/developer/providers/ipam.md]: I propose to merge this as is, and we'll update it if we change how this works in the in-cluster ipam implementation.

**Comment by fabriziopandini** [comment]
Thanks for documenting this contract, good documentation is the foundation of CAPI extensibility!
/lgtm
/approve

**Pull request merged**